File: backend/src/main/java/com/group8/sams/service/AttendanceService.java
package com.group8.sams.service;

import com.group8.sams.dto.request.AttendanceRequest;
import com.group8.sams.dto.request.BulkAttendanceRequest;
import com.group8.sams.dto.response.AttendanceResponse;
import com.group8.sams.dto.response.AttendanceSummaryResponse;
import com.group8.sams.security.UserPrincipal;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

/**
 * Every method takes the caller, because attendance authorization is not just
 * about roles - it depends on which course the record belongs to and which student
 * it is about.
 */
public interface AttendanceService {

    AttendanceResponse record(AttendanceRequest request, UserPrincipal caller);

    List<AttendanceResponse> recordBulk(BulkAttendanceRequest request, UserPrincipal caller);

    AttendanceResponse update(Long id, AttendanceRequest request, UserPrincipal caller);

    void delete(Long id, UserPrincipal caller);

    List<AttendanceResponse> findByCourseAndDate(Long courseId, LocalDate date,
                                                 UserPrincipal caller);

    List<AttendanceResponse> findByStudent(Long studentId, UserPrincipal caller);

    AttendanceSummaryResponse percentage(Long studentId, Long courseId, UserPrincipal caller);

    List<AttendanceSummaryResponse> courseSummary(Long courseId, UserPrincipal caller);

    List<AttendanceSummaryResponse> lowAttendance(BigDecimal threshold, UserPrincipal caller);
}

File: frontend/src/services/attendanceService.js
import api from './api';

/** Attendance API. Owner: Member 3. */
const attendanceService = {
  record: (payload) => api.post('/attendance', payload).then((r) => r.data),
  recordBulk: (payload) => api.post('/attendance/bulk', payload).then((r) => r.data),
  update: (id, payload) => api.put(`/attendance/${id}`, payload).then((r) => r.data),
  remove: (id) => api.delete(`/attendance/${id}`),
  byCourse: (courseId, date) =>
    api.get(`/attendance/course/${courseId}`, { params: date ? { date } : {} })
      .then((r) => r.data),
  byStudent: (studentId) => api.get(`/attendance/student/${studentId}`).then((r) => r.data),
  percentage: (studentId, courseId) =>
    api.get('/attendance/percentage', { params: { studentId, courseId } }).then((r) => r.data),
  courseSummary: (courseId) =>
    api.get(`/attendance/summary/course/${courseId}`).then((r) => r.data),
  low: (threshold) =>
    api.get('/attendance/low', { params: threshold ? { threshold } : {} }).then((r) => r.data),
};

export default attendanceService;
